contracts/index_executor/src/storage.rs

```
@@ -0,0 +1,323 @@
+        let access_key = format!("Bearer {}", self.key);
+        let headers: Vec<(String, String)> = vec![
+            ("Content-Type".into(), "application/json".into()),
+            ("Authorization".into(), access_key),
+            ("Content-Length".into(), content_length),
+        ];
+
+        let response: pink_extension::chain_extension::HttpResponse = http_req!(
+            method,
+            format!("{}{}", self.url.clone(), api),
+            request.as_bytes().to_vec(),
+            headers
+        );
+        if response.status_code != 200 {
+            return Err("CallServiceFailed");
+        }
+
+        Ok(response.body)
+    }
+
+    /// Return (encoded_data, document_id) if success
+    fn read_storage(&self, key: &[u8]) -> Result<Option<(Vec<u8>, String)>, &'static str> {
+        let key = key
+            .iter()
+            .map(|byte| format!("{byte:02x}"))
+            .collect::<String>();
+        pink_extension::debug!("read_storage: id: {}", key);
+
+        let cmd = format!(
+            r#"{{
+                "structuredQuery": {{
+                    "from": [{{
+                      "collectionId": "index-storage"
+                    }}],
+                    "where": {{
+                      "fieldFilter": {{
+                        "field": {{
+                          "fieldPath": "id"
+                        }},
+                        "op": "EQUAL",
+                        "value": {{
+                          "stringValue": "{key}"
+                        }}
+                      }}
+                    }},
+                    "limit": 1
+                  }}
+            }}"#
+        );
+
+        let response_body: Vec<u8> = self.send_request("POST", "documents:runQuery", &cmd)?;
+        if let Ok(response) = pink_json::from_slice::<Vec<ResponseData>>(&response_body) {
+            Ok(if !response.is_empty() {
+                let data_str = response[0].document.fields.data.string_value.clone();
```

Please follow the "return early" style error handling. So instead of write if like this:

```
if ok1 {
  if ok2 {
    return Ok()
  } else {
    return Err()
  }
} else {
  if ok3 {
    return Ok()
  } else {
    return Err()
  }
}
```

You can return the error earlier to make the indent level less:

```
if !ok1 {
  if !ok3 {
    return Err()
  }
  return Ok()
}
if !ok2 {
  return Err()
}
Ok()
```

Reference: https://medium.com/swlh/return-early-pattern-3d18a41bba8

---

Done